GVFS/GVFS.Common/Cleanup/PrefetchStep.cs

```diff
@@ -1,105 +1,147 @@
             }
 
-            if (packIndexes?.Count > 0)
+            this.TrySchedulePostFetchJob(packIndexes);
+
+            return true;
+        }
+
+        protected override void RunGitAction()
+        {
+            long last;
+            string error = null;
+
+            if (!this.TryGetMaxGoodPrefetchTimestamp(out last, out error))
             {
-                TrySchedulePostFetchJob(tracer, enlistment.NamedPipeName, packIndexes);
+                this.Context.Tracer.RelatedError(error);
+                return;
             }
 
-            return true;
+            DateTime lastDateTime = EpochConverter.FromUnixEpochSeconds(last);
+            DateTime now = DateTime.UtcNow;
+
+            if (now <= lastDateTime + this.timeBetweenPrefetches)
+            {
+                this.Context.Tracer.RelatedInfo(this.TelemetryKey + ": Skipping prefetch since most-recent prefetch ({0}) is too close to now ({1})", lastDateTime, now);
+                return;
+            }
+
+            this.RunGitCommand(process =>
+            {
+                this.TryPrefetchCommitsAndTrees(out error, process);
+                return null;
+            });
+
+            if (!string.IsNullOrEmpty(error))
+            {
+                this.Context.Tracer.RelatedWarning(
+                    metadata: null,
+                    message: $"{this.TelemetryKey}: {nameof(this.TryPrefetchCommitsAndTrees)} failed with error '{error}'",
+                    keywords: Keywords.Telemetry);
+            }
         }
 
-        public static bool TryGetMaxGoodPrefetchTimestamp(
-            ITracer tracer,
-            GVFSEnlistment enlistment,
-            PhysicalFileSystem fileSystem,
-            GitObjects gitObjects,
-            out long maxGoodTimestamp,
-            out string error)
+        private bool TryGetMaxGoodPrefetchTimestamp(out long maxGoodTimestamp, out string error)
         {
-            fileSystem.CreateDirectory(enlistment.GitPackRoot);
+            this.Context.FileSystem.CreateDirectory(this.Context.Enlistment.GitPackRoot);
 
-            string[] packs = gitObjects.ReadPackFileNames(enlistment.GitPackRoot, GVFSConstants.PrefetchPackPrefix);
+            string[] packs = GitObjects.ReadPackFileNames(this.Context.Enlistment.GitPackRoot, GVFSConstants.PrefetchPackPrefix);
             List<PrefetchPackInfo> orderedPacks = packs
-                .Where(pack => GetTimestamp(pack).HasValue)
-                .Select(pack => new PrefetchPackInfo(GetTimestamp(pack).Value, pack))
+                .Where(pack => this.GetTimestamp(pack).HasValue)
+                .Select(pack => new PrefetchPackInfo(this.GetTimestamp(pack).Value, pack))
                 .OrderBy(packInfo => packInfo.Timestamp)
                 .ToList();
 
             maxGoodTimestamp = -1;
 
             int firstBadPack = -1;
-            for (int i = 0; i < orderedPacks.Count; ++i)
+            for (int i = 0; !this.Stopping && i < orderedPacks.Count; ++i)

```

Comment:
We check for this.Stopping a lot in the step.  Wondering if it's worth just letting the kill process do it's job so it isn't a frequent check in the step.